---
output: html_notebook
---

```{r}
library('tidyverse')
library('readr')
library('tidyr')
cols = c('chrom', 'source', 'feature_type', 'start', 'end', 'score', 'strand', 'phase', 'info')
d <- read_delim('gencode.v31.basic.annotation.gtf.gz', 
               delim='\t', skip = 5, progress = F, 
               col_names = cols)
```

# Task
1. Computes the number of transcripts per gene. What is the mean number of transcripts per gene? What is the quantile (25%, 50%, 75%) for these numbers? Which gene has the greatest number of transcript?  

```{r}
d1 <- d %>%
  filter(feature_type == c('gene', 'transcript'))
```

```{r}
d2 <- d1 %>%
  select(chrom, feature_type, start, end) %>%
  arrange(chrom, start, feature_type)
```

```{r}
d_gene <- d2 %>%
  filter(feature_type == 'gene') %>%
  mutate(genenumber = order(chrom, start))
```

```{r}
d_transcript <- d2 %>%
  filter(feature_type == 'transcript')
```

```{r}
d_transcript2 <- merge(d_transcript[,], d_gene[, c("chrom","start", "end","genenumber")], all.x = TRUE)
```

```{r}
d_numbered <- rbind(d_gene, d_transcript2) %>%
  arrange(chrom, start, feature_type)
```

```{r}
d_numbered <- fill(d_numbered, genenumber)
```

```{r}
transcripts_per_gene <- d_numbered %>%
  filter(feature_type == 'transcript') %>%
  group_by(genenumber) %>%
  tally()
```

```{r}
transcripts_per_gene %>%
  group_by(n) %>%
  tally()
```


```{r}
transcripts_per_gene %>%
  summarize(average = mean(n))
```

```{r}
my_quantile <- function(dat){
  x <- quantile(dat$n, c(0.25, 0.5, 0.75))
  tibble(quarter = x[1], half = x[2], three_quarter = x[3])
}
transcripts_per_gene %>%
  do(my_quantile(.))
```

```{r}
i_max <- which.max(transcripts_per_gene$n)
transcripts_per_gene$genenumber[i_max]
```


2. Compute the number of transcripts per gene among gene biotypes. For example, compare the number of transcript per gene between protein-coding genes, long noncoding genes, pseudogenes. You can find the details of gene biotype here. https://www.gencodegenes.org/pages/biotypes.html  


3. Final task is to compute the number of transcripts per gene per chromosome.  

```{r}
transcripts_per_gene_per_chromosome <-
  d_numbered %>%
  filter(feature_type == 'transcript') %>%
  group_by(chrom, genenumber) %>%
  tally()
```

```{r}
average_per_gene_per_chromosome <- transcripts_per_gene_per_chromosome %>%
  summarize(average = mean(n))
average_per_gene_per_chromosome
```


